op-node/rollup/derive/engine_queue.go

```diff
@@ -427,6 +430,20 @@ func (eq *EngineQueue) tryNextUnsafePayload(ctx context.Context) error {
 }
 
 func (eq *EngineQueue) tryNextSafeAttributes(ctx context.Context) error {
+	if eq.safeAttributes == nil { // sanity check the attributes are there
+		return nil
+	}
+	// validate the safe attributes before processing them. The engine may have completed processing them through other means.
+	if eq.safeHead != eq.safeAttributesParent {
+		if eq.safeHead.ParentHash != eq.safeAttributesParent.Hash {
```

Comment:
won't we always hit this case & force a reset if `eq.safeHead != eq.safeAttributesParent` is true? My understanding of the code is that if the safeHead has been updated separately from this, we want to cancel this safeAttributes, but we also take this chance to check for a L2 reorg.

I'm also slightly worried about cancelling a safeAttributes here.

Reply:
See PR description, we should have some sanity check that the attributes we have are consistent with the state that we generated them for. If that state isn't consistent anymore, and we didn't reset, it means the engine went ahead and continued to process something that was attempted last. And it'll allow us to separate the engine and the derivation more.